File: ios/ArViewerViewManager.m
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(ArViewerViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(model, NSString)
RCT_EXPORT_VIEW_PROPERTY(planeOrientation, NSString)
RCT_EXPORT_VIEW_PROPERTY(allowScale, BOOL)
RCT_EXPORT_VIEW_PROPERTY(allowRotate, BOOL)
RCT_EXPORT_VIEW_PROPERTY(allowTranslate, BOOL)
RCT_EXPORT_VIEW_PROPERTY(lightEstimation, BOOL)
RCT_EXPORT_VIEW_PROPERTY(manageDepth, BOOL)
RCT_EXPORT_VIEW_PROPERTY(disableInstructions, BOOL)
RCT_EXPORT_VIEW_PROPERTY(disableInstantPlacement, BOOL)

RCT_EXTERN_METHOD(reset:(nonnull NSNumber*) reactTag)
RCT_EXTERN_METHOD(loadModel:(nonnull NSNumber*) reactTag)
RCT_EXTERN_METHOD(takeScreenshot:(nonnull NSNumber*)reactTag withRequestId:(nonnull NSNumber*)requestId)
RCT_EXTERN_METHOD(getPositionVector3:(nonnull NSNumber*)reactTag withXCoord:(nonnull NSNumber*)x withYCoord:(nonnull NSNumber*)y withRequestId:(nonnull NSNumber*)requestId)
RCT_EXTERN_METHOD(createLineAndGetDistance:(nonnull NSNumber*)reactTag withX1Coord:(nonnull NSNumber*)x1 withY1Coord:(nonnull NSNumber*)y1 withZ1Coord:(nonnull NSNumber*)z1 withX2Coord:(nonnull NSNumber*)x2 withY2Coord:(nonnull NSNumber*)y2 withZ2Coord:(nonnull NSNumber*)z2 withColor:(nonnull NSString*)color withRequestId:(nonnull NSNumber*)requestId)
RCT_EXTERN_METHOD(placeModel:(nonnull NSNumber*)reactTag withXCoord:(nonnull NSNumber*)x withYCoord:(nonnull NSNumber*)y withZCoord:(nonnull NSNumber*)z)
RCT_EXTERN_METHOD(placeText:(nonnull NSNumber*)reactTag withXCoord:(nonnull NSNumber*)x withYCoord:(nonnull NSNumber*)y withZCoord:(nonnull NSNumber*)z withColor:(nonnull NSString*)color withText:(nonnull NSString*)text)
RCT_EXTERN_METHOD(rotateModel:(nonnull NSNumber*)reactTag withPitch:(nonnull NSNumber*)pitch withYaw:(nonnull NSNumber*)yaw withRoll:(nonnull NSNumber*)roll)

RCT_EXPORT_VIEW_PROPERTY(onStarted, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onUserTap, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onDataReturned, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onEnded, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onModelPlaced, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onModelRemoved, RCTDirectEventBlock)

@end

File: ios/ArViewerViewManager.swift
import ARKit
import RealityKit
import SwiftUI

@available(iOS 13.0, *)
@objc(ArViewerViewManager)
class ArViewerViewManager: RCTViewManager {
    @objc override static func requiresMainQueueSetup() -> Bool {
        return false
    }
        
    override func view() -> UIView {
        return ArViewerView()
    }
    
    @objc func reset(_ node : NSNumber){
        RCTExecuteOnMainQueue {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? ArViewerView {
                view.reset()
            }
        }
    }
    
    @objc func loadModel(_ node : NSNumber){
        RCTExecuteOnMainQueue {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? ArViewerView {
                view.loadModel()
            }
        }
    }
    
    @objc func takeScreenshot(_ node : NSNumber, withRequestId requestId: NSNumber){
        RCTExecuteOnMainQueue {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? ArViewerView {
                view.takeScreenshot(requestId: requestId.intValue)
            }
        }
    }
    
    @objc func getPositionVector3(_ node : NSNumber,withXCoord x: NSNumber, withYCoord y: NSNumber, withRequestId requestId: NSNumber){
        RCTExecuteOnMainQueue {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? ArViewerView {
                view.getPositionVector3(x: x.doubleValue,y: y.doubleValue, requestId: requestId.intValue)
            }
        }
    }
    
    @objc func createLineAndGetDistance(_ node : NSNumber,withX1Coord x1: NSNumber, withY1Coord y1: NSNumber, withZ1Coord z1: NSNumber,withX2Coord x2: NSNumber, withY2Coord y2: NSNumber, withZ2Coord z2: NSNumber,withColor color:NSString, withRequestId requestId: NSNumber){
        RCTExecuteOnMainQueue {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? ArViewerView {
                view.createLineAndGetDistance(x1: x1.doubleValue,y1: y1.doubleValue,z1: z1.doubleValue,x2: x2.doubleValue,y2: y2.doubleValue,z2: z2.doubleValue,color: color as String, requestId: requestId.intValue)
            }
        }
    }
    
    @objc func placeText(_ node : NSNumber,withXCoord x: NSNumber, withYCoord y: NSNumber, withZCoord z: NSNumber,withColor color:NSString, withText text: NSString){
        RCTExecuteOnMainQueue {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? ArViewerView {
                view.placeText(x: x.doubleValue,y: y.doubleValue, z: z.doubleValue,color: color as String,text: text as String)
            }
        }
    }
    
    @objc func placeModel(_ node : NSNumber,withXCoord x: NSNumber, withYCoord y: NSNumber, withZCoord z: NSNumber){
        RCTExecuteOnMainQueue {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? ArViewerView {
                view.placeModel(x: x.doubleValue,y: y.doubleValue, z: z.doubleValue)
            }
        }
    }
    
    @objc func rotateModel(_ node : NSNumber, withPitch pitch: NSNumber, withYaw yaw: NSNumber, withRoll roll: NSNumber){
        RCTExecuteOnMainQueue {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? ArViewerView {
                view.rotateModel(pitch: pitch.intValue, yaw: yaw.intValue, roll: roll.intValue)
            }
        }
    }
}


extension UIView {
    var parentViewController: UIViewController? {
        var parentResponder: UIResponder? = self
        while parentResponder != nil {
            parentResponder = parentResponder!.next
            if let viewController = parentResponder as? UIViewController {
                return viewController
            }
        }
        return nil
    }
}
